File: Advanced DS/K-D Trees/BasicKDTree.cpp
/**
 * Basic K-D Tree Implementation
 * 
 * Standard K-D Tree with sequential insertion.
 * Simple structure cycling through dimensions at each level.
 */

#include <iostream>
#include <vector>
#include <memory>
#include <cmath>
#include <limits>
#include <algorithm>
using namespace std;

// Point in K-dimensional space
template<int K>
struct Point {
    double coords[K];
    string data;
    
    Point() : data("") {
        for (int i = 0; i < K; i++) coords[i] = 0;
    }
    
    Point(const vector<double>& c, const string& d = "") : data(d) {
        for (int i = 0; i < K && i < c.size(); i++) {
            coords[i] = c[i];
        }
    }
    
    double operator[](int index) const {
        return coords[index];
    }
    
    double& operator[](int index) {
        return coords[index];
    }
    
    // Euclidean distance to another point
    double distanceTo(const Point& other) const {
        double sum = 0;
        for (int i = 0; i < K; i++) {
            double diff = coords[i] - other.coords[i];
            sum += diff * diff;
        }
        return sqrt(sum);
    }
    
    void print() const {
        cout << "(";
        for (int i = 0; i < K; i++) {
            cout << coords[i];
            if (i < K - 1) cout << ", ";
        }
        cout << ")";
        if (!data.empty()) cout << " - " << data;
    }
};

// K-D Tree Node
template<int K>
class KDTreeNode {
public:
    Point<K> point;
    unique_ptr<KDTreeNode> left;
    unique_ptr<KDTreeNode> right;
    
    KDTreeNode(const Point<K>& p) : point(p), left(nullptr), right(nullptr) {}
};

// Basic K-D Tree
template<int K>
class BasicKDTree {
private:
    unique_ptr<KDTreeNode<K>> root;
    
    // Insert helper
    void insertHelper(unique_ptr<KDTreeNode<K>>& node, const Point<K>& point, int depth) {
        if (!node) {
            node = make_unique<KDTreeNode<K>>(point);
            return;
        }
        
        int axis = depth % K;
        
        if (point[axis] < node->point[axis]) {
            insertHelper(node->left, point, depth + 1);
        } else {
            insertHelper(node->right, point, depth + 1);
        }
    }
    
    // Search helper
    bool searchHelper(KDTreeNode<K>* node, const Point<K>& point, int depth) {
        if (!node) return false;
        
        // Check if this is the point
        bool match = true;
        for (int i = 0; i < K; i++) {
            if (node->point[i] != point[i]) {
                match = false;
                break;
            }
        }
        if (match) return true;
        
        int axis = depth % K;
        
        if (point[axis] < node->point[axis]) {
            return searchHelper(node->left.get(), point, depth + 1);
        } else {
            return searchHelper(node->right.get(), point, depth + 1);
        }
    }
    
    // Nearest neighbor helper
    void nearestHelper(KDTreeNode<K>* node, const Point<K>& target,
                      Point<K>*& best, double& bestDist, int depth) {
        if (!node) return;
        
        double dist = node->point.distanceTo(target);
        if (dist < bestDist) {
            bestDist = dist;
            best = &node->point;
        }
        
        int axis = depth % K;
        double diff = target[axis] - node->point[axis];
        
        // Choose which subtree to search first
        KDTreeNode<K>* nearSubtree = (diff < 0) ? node->left.get() : node->right.get();
        KDTreeNode<K>* farSubtree = (diff < 0) ? node->right.get() : node->left.get();
        
        // Search near subtree
        nearestHelper(nearSubtree, target, best, bestDist, depth + 1);
        
        // Check if we need to search far subtree
        if (abs(diff) < bestDist) {
            nearestHelper(farSubtree, target, best, bestDist, depth + 1);
        }
    }
    
    // Range query helper
    void rangeQueryHelper(KDTreeNode<K>* node, const Point<K>& lower,
                         const Point<K>& upper, vector<Point<K>>& results, int depth) {
        if (!node) return;
        
        // Check if point is in range
        bool inRange = true;
        for (int i = 0; i < K; i++) {
            if (node->point[i] < lower[i] || node->point[i] > upper[i]) {
                inRange = false;
                break;
            }
        }
        if (inRange) {
            results.push_back(node->point);
        }
        
        int axis = depth % K;
        
        // Check left subtree if necessary
        if (lower[axis] <= node->point[axis]) {
            rangeQueryHelper(node->left.get(), lower, upper, results, depth + 1);
        }
        
        // Check right subtree if necessary
        if (upper[axis] >= node->point[axis]) {
            rangeQueryHelper(node->right.get(), lower, upper, results, depth + 1);
        }
    }
    
    // Count nodes
    int countNodes(KDTreeNode<K>* node) {
        if (!node) return 0;
        return 1 + countNodes(node->left.get()) + countNodes(node->right.get());
    }
    
    // Calculate height
    int getHeight(KDTreeNode<K>* node) {
        if (!node) return 0;
        return 1 + max(getHeight(node->left.get()), getHeight(node->right.get()));
    }
    
public:
    BasicKDTree() : root(nullptr) {}
    
    // Insert a point
    void insert(const Point<K>& point) {
        insertHelper(root, point, 0);
    }
    
    // Search for a point
    bool search(const Point<K>& point) {
        return searchHelper(root.get(), point, 0);
    }
    
    // Find nearest neighbor
    Point<K>* nearestNeighbor(const Point<K>& target) {
        Point<K>* best = nullptr;
        double bestDist = numeric_limits<double>::max();
        nearestHelper(root.get(), target, best, bestDist, 0);
        return best;
    }
    
    // Range query
    vector<Point<K>> rangeQuery(const Point<K>& lower, const Point<K>& upper) {
        vector<Point<K>> results;
        rangeQueryHelper(root.get(), lower, upper, results, 0);
        return results;
    }
    
    // Print statistics
    void printStats() {
        int nodes = countNodes(root.get());
        int height = getHeight(root.get());
        cout << "Nodes: " << nodes << ", Height: " << height << endl;
    }
};

// Example usage
int main() {
    cout << "=== Basic 2D K-D Tree Demo ===" << endl << endl;
    
    BasicKDTree<2> kdtree;
    
    // Insert points
    vector<Point<2>> points = {
        Point<2>({7, 2}, "A"),
        Point<2>({5, 4}, "B"),
        Point<2>({9, 6}, "C"),
        Point<2>({2, 3}, "D"),
        Point<2>({4, 7}, "E"),
        Point<2>({8, 1}, "F")
    };
    
    cout << "Inserting points:" << endl;
    for (const auto& p : points) {
        kdtree.insert(p);
        cout << "  ";
        p.print();
        cout << endl;
    }
    
    // Print stats
    cout << "\nTree Statistics:" << endl;
    kdtree.printStats();
    
    // Search for a point
    cout << "\nSearching for point (5, 4):" << endl;
    Point<2> searchPoint({5, 4});
    bool found = kdtree.search(searchPoint);
    cout << "  " << (found ? "Found!" : "Not found") << endl;
    
    // Nearest neighbor
    cout << "\nNearest neighbor to (6, 3):" << endl;
    Point<2> target({6, 3});
    Point<2>* nearest = kdtree.nearestNeighbor(target);
    if (nearest) {
        cout << "  ";
        nearest->print();
        cout << " at distance " << target.distanceTo(*nearest) << endl;
    }
    
    // Range query
    cout << "\nRange query [3, 8] x [2, 6]:" << endl;
    Point<2> lower({3, 2});
    Point<2> upper({8, 6});
    vector<Point<2>> results = kdtree.rangeQuery(lower, upper);
    cout << "  Found " << results.size() << " points:" << endl;
    for (const auto& p : results) {
        cout << "    ";
        p.print();
        cout << endl;
    }
    
    return 0;
}

File: Level 6 - Specialist/K-D Trees/BasicKDTree.java
/**
 * Basic K-D Tree Implementation in Java
 * Standard K-D Tree with sequential insertion
 */

import java.util.*;

class Point {
    double[] coords;
    String data;
    
    public Point(double[] coords, String data) {
        this.coords = coords.clone();
        this.data = data;
    }
    
    public int dimensions() {
        return coords.length;
    }
    
    public double distanceTo(Point other) {
        double sum = 0;
        for (int i = 0; i < coords.length; i++) {
            double diff = coords[i] - other.coords[i];
            sum += diff * diff;
        }
        return Math.sqrt(sum);
    }
    
    @Override
    public String toString() {
        return Arrays.toString(coords) + (data != null ? " - " + data : "");
    }
}

class KDTreeNode {
    Point point;
    KDTreeNode left, right;
    
    public KDTreeNode(Point point) {
        this.point = point;
    }
}

public class BasicKDTree {
    private KDTreeNode root;
    private int k;
    
    public BasicKDTree(int k) {
        this.k = k;
    }
    
    public void insert(Point point) {
        root = insertHelper(root, point, 0);
    }
    
    private KDTreeNode insertHelper(KDTreeNode node, Point point, int depth) {
        if (node == null) {
            return new KDTreeNode(point);
        }
        
        int axis = depth % k;
        
        if (point.coords[axis] < node.point.coords[axis]) {
            node.left = insertHelper(node.left, point, depth + 1);
        } else {
            node.right = insertHelper(node.right, point, depth + 1);
        }
        
        return node;
    }
    
    public Point nearestNeighbor(Point target) {
        if (root == null) return null;
        
        Point[] best = new Point[]{null};
        double[] bestDist = new double[]{Double.MAX_VALUE};
        
        nearestHelper(root, target, best, bestDist, 0);
        
        return best[0];
    }
    
    private void nearestHelper(KDTreeNode node, Point target, 
                              Point[] best, double[] bestDist, int depth) {
        if (node == null) return;
        
        double dist = node.point.distanceTo(target);
        if (dist < bestDist[0]) {
            bestDist[0] = dist;
            best[0] = node.point;
        }
        
        int axis = depth % k;
        double diff = target.coords[axis] - node.point.coords[axis];
        
        KDTreeNode nearSubtree = (diff < 0) ? node.left : node.right;
        KDTreeNode farSubtree = (diff < 0) ? node.right : node.left;
        
        nearestHelper(nearSubtree, target, best, bestDist, depth + 1);
        
        if (Math.abs(diff) < bestDist[0]) {
            nearestHelper(farSubtree, target, best, bestDist, depth + 1);
        }
    }
    
    public List<Point> rangeQuery(Point lower, Point upper) {
        List<Point> results = new ArrayList<>();
        rangeQueryHelper(root, lower, upper, results, 0);
        return results;
    }
    
    private void rangeQueryHelper(KDTreeNode node, Point lower, Point upper,
                                  List<Point> results, int depth) {
        if (node == null) return;
        
        boolean inRange = true;
        for (int i = 0; i < k; i++) {
            if (node.point.coords[i] < lower.coords[i] || 
                node.point.coords[i] > upper.coords[i]) {
                inRange = false;
                break;
            }
        }
        if (inRange) {
            results.add(node.point);
        }
        
        int axis = depth % k;
        
        if (lower.coords[axis] <= node.point.coords[axis]) {
            rangeQueryHelper(node.left, lower, upper, results, depth + 1);
        }
        if (upper.coords[axis] >= node.point.coords[axis]) {
            rangeQueryHelper(node.right, lower, upper, results, depth + 1);
        }
    }
    
    public static void main(String[] args) {
        System.out.println("=== Basic K-D Tree Demo ===\n");
        
        BasicKDTree kdtree = new BasicKDTree(2);
        
        Point[] points = {
            new Point(new double[]{7, 2}, "A"),
            new Point(new double[]{5, 4}, "B"),
            new Point(new double[]{9, 6}, "C"),
            new Point(new double[]{2, 3}, "D"),
            new Point(new double[]{4, 7}, "E"),
            new Point(new double[]{8, 1}, "F")
        };
        
        System.out.println("Inserting points:");
        for (Point p : points) {
            kdtree.insert(p);
            System.out.println("  " + p);
        }
        
        System.out.println("\nNearest neighbor to (6, 3):");
        Point target = new Point(new double[]{6, 3}, null);
        Point nearest = kdtree.nearestNeighbor(target);
        if (nearest != null) {
            System.out.println("  " + nearest + " at distance " + 
                              target.distanceTo(nearest));
        }
        
        System.out.println("\nRange query [3, 8] x [2, 6]:");
        Point lower = new Point(new double[]{3, 2}, null);
        Point upper = new Point(new double[]{8, 6}, null);
        List<Point> results = kdtree.rangeQuery(lower, upper);
        System.out.println("  Found " + results.size() + " points:");
        for (Point p : results) {
            System.out.println("    " + p);
        }
    }
}
